## Model Training
### Tasks including -
- **Train a Model:** Use a dataset (like ```MNIST```) to train your 3-layer deep network.
- **Save the Model:** Save the trained model to a file.
- **Test the Model Accuracy:** Test the accuracy of the Model.

<hr>

### Importing the Libraries

In [3]:
import os
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Input
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

### Handle GPU Issues (Optional)
This step is in response of error indicates a problem with the GPU setup, specifically with CUDA, which TensorFlow uses for GPU acceleration. The error ```UNKNOWN ERROR (303)``` suggests that TensorFlow is unable to initialize CUDA. To solve this issue this step is initialized here. __*It's Optional!*__

In [5]:
# If CUDA errors persist, force TensorFlow to use the CPU
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"  # Comment this out if GPU setup is fixed

# Verify if TensorFlow is using the correct device
print("TensorFlow is using:", tf.config.list_physical_devices('GPU') or "CPU")

TensorFlow is using: CPU


### Load and Preprocess MNIST Data

In [7]:
# MNIST data loading
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalize pixel values to [0, 1]
X_train = X_train / 255.0
X_test = X_test / 255.0

# Reshape data to include the channel dimension (samples, 28, 28, 1)
X_train = X_train.reshape(-1, 28, 28, 1)
X_test = X_test.reshape(-1, 28, 28, 1)

# Convert labels to one-hot encoding
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

### Define the 3-Layer Deep Neural Network

In [8]:
model = Sequential([
    Input(shape=(28, 28, 1)),       # Explicit Input layer
    Flatten(),                      # Flatten input images to 1D arrays
    Dense(128, activation='relu'),  # First hidden layer with 128 neurons
    Dense(64, activation='relu'),   # Second hidden layer with 64 neurons
    Dense(10, activation='softmax') # Output layer with 10 neurons for digits 0-9
])

### Compile the Model

In [9]:
model.compile(optimizer='adam',                  # Adam optimizer
              loss='categorical_crossentropy',   # Loss function for multi-class classification
              metrics=['accuracy'])              # Evaluate accuracy during training


### Train the Model

In [10]:
history = model.fit(X_train, y_train,
                    epochs=10,                  # Train for 10 epochs
                    batch_size=32,              # Use a batch size of 32
                    validation_split=0.2)       # Use 20% of training data for validation

2024-11-21 17:43:17.468989: W external/local_xla/xla/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 150528000 exceeds 10% of free system memory.


Epoch 1/10
1486/1500 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8633 - loss: 0.4670

2024-11-21 17:43:21.924396: W external/local_xla/xla/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 37632000 exceeds 10% of free system memory.


1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8639 - loss: 0.4650 - val_accuracy: 0.9597 - val_loss: 0.1358
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9657 - loss: 0.1164 - val_accuracy: 0.9654 - val_loss: 0.1137
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9775 - loss: 0.0748 - val_accuracy: 0.9685 - val_loss: 0.1045
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9829 - loss: 0.0541 - val_accuracy: 0.9728 - val_loss: 0.0955
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9853 - loss: 0.0451 - val_accuracy: 0.9693 - val_loss: 0.1027
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9886 - loss: 0.0348 - val_accuracy: 0.9732 - val_loss: 0.1024
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9915 - loss: 0.0281 - val_accuracy: 0.9743 - val_loss: 0.1048
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9925 - loss: 0.0244 - val_accurac

### Save the Trained Model

In [12]:
model.save("my_trained_model.h5")
print("Model saved successfully as 'my_trained_model.h5'")

Model saved successfully as 'my_trained_model.h5'


### Evaluate the Model

In [13]:
test_loss, test_accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f"Test Accuracy: {test_accuracy * 100:.2f}%")

2024-11-21 17:44:39.913774: W external/local_xla/xla/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 31360000 exceeds 10% of free system memory.


Test Accuracy: 97.88%


___

### Instructions to Use the Model
The model is successfully trained with the 97.88% accuracy. Now, use the Model for Your Handwritten Digits.
After verifying the model, you can use it for your handwritten digits:

- Load the saved model:
```python
from tensorflow.keras.models import load_model

# Load the saved model
model = load_model("my_trained_model.h5")
```
- Preprocess your handwritten images.
- Use the model for predictions.

__*Thank you!*__